Automattic/wp-calypso
# [Highly Experimental] Theme Showcase: Refactor the Theme Card

This is an **experimental refactor of the Calypso `Theme` component** used by the Theme Showcase (TS) and Design Picker (DP).
 
<img width="200" alt="Screenshot 2023-11-10 at 14 50 53" src="https://github.com/Automattic/wp-calypso/assets/2070010/328ec999-130a-42f8-916d-07516354d1bf">

The TS is an old section of Calypso that has sustained many updates, patches, workarounds, and hacks over the years. Since it's at the center of many flows, it's one of Calypso's most complex parts, and one of the hardest to modernize for fear of regressing unexpectedly on unknown flows.

I've tried a few times to refactor it all, but it's a daunting task indeed. Eventually, I decided to start from the last leaf of the component tree: the _humble_ `Theme` card.

Working on this PR gave me invaluable knowledge of the inner workings of the TS, appreciating architectural decisions based on ancient React versions (and maybe even before Redux came into play!), and grumbling about shortcuts taken more recently that skyrocketed the complexity either for time constraints or for plain old inexperience.

## The Technicalities

To avoid destroying literally everything, I've decided to develop the new `Theme` component in the [`themenew`](https://github.com/Automattic/wp-calypso/tree/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/components/themenew) folder, and use it behind a `themes/new-theme-card` feature flag, enabled by default on dev and wpcalypso.

The reason for this is to not affect production and to be of as little impediment as possible to the regular TS development, while also putting the new component front and center, hopefully catching bugs and regressions.

I have also introduced a few contexts used by the new `Theme` component. They are applied without feature flag since the current TS does not use them.

To centralize the development, all contexts and hooks are inside the same `themenew` folder. They should not be there and will be moved to their final location when (if!) this PR is about to be approved.

Lastly, I've decided to not touch the DP for now. I'm fairly sure the `Theme` card proposed here doesn't work there, but I'm also convinced that the development experience improvements will make adapting it for the DP relatively easy.

### Before: The Theme Options

Arguably, the worst part of the `Theme` card is the [`theme-options.js`](https://github.com/Automattic/wp-calypso/blob/6c97adb1f0ce3ce0f95ef2199039b63a3c1acc3a/client/my-sites/themes/theme-options.js) file which controls the theme's _actions_, listed in the more menu.

Theme ~~options~~ actions are centralized in an object controlling labels, "hrefs" and "onClicks", and the visibility logic.
The object is [connected to Redux separately](https://github.com/Automattic/wp-calypso/blob/d9e3960e9d107ea39abc56e44e499f18d7f5c973/client/my-sites/themes/theme-options.js#L328), its methods are liberally passed around to gain access to the state, and its [actions are decorated](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/themes/themes-selection.jsx#L135) and its URLs [decorated](https://github.com/Automattic/wp-calypso/blob/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/my-sites/themes/helpers.js#L73-L81) and [localized](https://github.com/Automattic/wp-calypso/blob/d9e3960e9d107ea39abc56e44e499f18d7f5c973/client/my-sites/themes/theme-options.js#L334-L335) elsewhere, [tracking is injected](https://github.com/Automattic/wp-calypso/blob/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/my-sites/themes/helpers.js#L7-L25) instead of being handled transparently. It's **impossible**, during our regular job, to fully understand all the implications of theme options. It takes an exploratory spike like this — and I can't encourage anyone to do it during their work hours.

The absolute amount of antipatterns and self-inflicted complications of `theme-options.js` is single-handedly responsible for making the TS finicky to work with and extend sanely.

Every developer working on the TS will agree with me that simple tasks that should take 1 hour, can easily take 1 week in the TS.

### After: The `ThemeActions`

I decided to replace the `theme-options.js` file with a [`ThemeAction`](https://github.com/Automattic/wp-calypso/blob/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/components/themenew/theme-actions/index.jsx) component which contains a single-responsibility child component for [each action](https://github.com/Automattic/wp-calypso/tree/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/components/themenew/theme-actions). Each action is then responsible for handling its own visibility logic, as well as its actions themselves.

All the action components contain a significant amount of `useSelectors`. I've asked opinions around, and I've eventually decided to keep it like this because it makes it for a better development experience without sacrificing performance: the popover menu itself is only rendered when visible, so these selectors won't be executed while the `Theme` is in a "resting state".

### Before: The Prop Drilling

Somewhat related to `theme-options.js` is the prop drilling that we do on the TS. Many props are carried from the routing controller all the way to the inner componentry of the `Theme` card.

This is obviously not a great architectural decision, which makes extending anything in the TS hard and bloated, forcing an unnecessarily tight coupling of all the components in play.

### After: The Context

This PR introduces three new contexts: [`ThemeContext`](https://github.com/Automattic/wp-calypso/blob/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/components/themenew/theme-context.js), [`ThemeCollectionContext`](https://github.com/Automattic/wp-calypso/blob/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/components/themenew/theme-collection-context.js), and [`ThemeShowcaseContext`](https://github.com/Automattic/wp-calypso/blob/9e54e238ed8a8c44c033c7b27766d84241c9c12a/client/components/themenew/theme-showcase-context.js).

They help reduce the prop drilling between TS components, decoupling them so that they can live independently from each other as long as they are used within the relevant contexts.

Practically speaking, now a `Theme` only needs three props:
- `themeId`, obviously.
- `themePosition`, the position in the list, for tracking purposes.
- `isPlaceholder`, to render a placeholder component instead.

Everything else is provided via context.

## What's Next?

If this PR makes sense to folks, then...

1. Replace the current theme card with the new `Theme`.
2. Move hooks and contexts where appropriate.
3. Eradicate all the theme options from everywhere under the sun to finally stop them from multiplying, and delete the `useDeprecatedThemeOptions` bridge, which is only there to support `ThemePreview` (I think?).
4. Remove a ton of stale code once and for all.

## Testing Instructions

- Run this PR locally, or use the Calypso live link.
- Open the Theme Showcase, either logged-in or logged-out.
- Check that the `Theme` cards are behaving as expected.
